File: Milestone 6/benim-portfolio-springboot-api/src/main/java/gcu/edu/Benim/Portfolio/Controllers/PortfolioController.java
package gcu.edu.Benim.Portfolio.Controllers;

import gcu.edu.Benim.Portfolio.Business.PortfolioService;
import gcu.edu.Benim.Portfolio.Data.JPAModels.Portfolio;
import gcu.edu.Benim.Portfolio.Models.DTO;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

/**
 *  Controller class for Portfolio model to implement all the REST-ful API actions
 */
@RestController
public class PortfolioController {
    @Autowired
    private PortfolioService service;
    private DTO dto;

    // Non-Default Constructor
    public PortfolioController(PortfolioService service){

        this.service = service;
    }

    // API GET call for {application:host}/portfolio to pull all the portfolio stored in database
    // returns the DTO model with the result data and necessary server-end information
    @GetMapping("/portfolios")
    DTO allPortfolios(){
        dto = new DTO(200, "Returned All portfolios!", this.service.getAllPortfolios());
        return dto;
    }

    // API POST call for {application:host}/portfolio to create a portfolio row in database
    // returns the DTO model with the result data and necessary server-end information
    @PostMapping("/portfolios")
    DTO newPortfolio(@RequestBody Portfolio newPortfolio){
        dto = new DTO(200, "Created portfolio with portfolio ID = " + newPortfolio.getPortfolioID(), this.service.createPortfolio(newPortfolio));
        return dto;
    }

    // API PUT call for {application:host}/portfolio/{id} to update the portfolio row in database with the given ID
    // returns the DTO model with the result data and necessary server-end information
    @PutMapping("/portfolios/{id}")
    DTO updatePortfolio(@RequestBody Portfolio updatedPortfolio, @PathVariable Long id){
        dto = new DTO(200, "Updated portfolio with ID = " + id , this.service.updatePortfolio(id, updatedPortfolio));
        return dto;
    }

    // API DELETE call for {application:host}/portfolio/{id} to delete the portfolio row in database with the given ID
    // returns the DTO model with the result data and necessary server-end information
    @DeleteMapping("/portfolios/{id}")
    DTO deletePortfolio(@PathVariable Long id){
        this.service.deletePortfolio(id);
        dto = new DTO(200, "Removed portfolio with ID = " + id, null);
        return dto;
    }

}

File: benim-portfolio-express-api/controllers/portfolio.controller.js
const service = require("../services/portfolio.service");
const {DTO} = require("../models/DTO");

/*
*   Calls the portfolio service to return all portfolio
*   then returns the result in DTO model as JSON
*/
exports.getPortfolios = (req, res, next) => {
  service.getPortfolios(req, (error, results) => {
    if (error) {
      return res.status(400).send({ success: 0, data: "Bad request" });
    }
    let dto = DTO(200, "Success", results)
    res.json(dto);
  });
};

/*
*   Calls the portfolio service to create the given portfolio
*   then returns the result in DTO model as JSON
*/
exports.createPortfolio = (req, res, next) => {
  service.createPortfolio(req.body, (error, results) => {
    if (error) {
      return res.status(400).send({ success: 0, data: "Bad Request" + error});
    }
    let dto = DTO(200, "Success", results)
    res.json(dto);
  });
};

/*
*   Calls the portfolio service to update a row of portfolio with the given ID
*   then returns the result in DTO model as JSON
*/
exports.updatePortfolio = (req, res, next) => {
  service.updatePortfolio(req, req.body, (error, results) => {
    if (error) {    
      return res.status(400).send({ success: 0, data: "Bad Request" + error});
    }
    let dto = DTO(200, "Success", results)
    res.json(dto);
  });
};

/*
*   Calls the portfolio service to delete a row of portfolio with the given ID
*   then returns the result in DTO model as JSON
*/
exports.deletePortfolio = (req, res, next) => {
  service.deletePortfolio(req, (error, results) => {
    if (error) {
      return res.status(400).send({ success: 0, data: "Bad Request" + error});
    }
    let dto = DTO(200, "Success", results)
    res.json(dto);
  });
};
